### PR TITLE
schedule reconnect in case of HTTP_GONE

### PR DESCRIPTION
**Pull request by honnix**
Reset resource version to null and schedule reconnect in case of HTTP_GHONE

This tries to address https://github.com/fabric8io/kubernetes-client/issues/1529

### CODE REVIEW

**Comment by honnix** [comment]
Couldn't figure out how this could fail `PATCH` request. 🤔

**Comment by rohanKanojia** [comment]
ok to test

**Comment by rohanKanojia** [comment]
@honnix : Not your fault. Our tests are flaky.

**Comment by honnix** [comment]
I seems I should fix these test cases as they look very relevant.

**Comment by rohanKanojia** [comment]
I meant integration tests, not unit tests

**Comment by honnix** [comment]
@rohanKanojia Yeah I'm working on it. Not very familiar with the code base, so I didn't find the test case.

File: kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/WatchTest.java [hunk 1]
```diff
@@ -58,38 +58,92 @@
...
+
+    final String path = "/api/v1/namespaces/test/pods?fieldSelector=metadata.name%3Dpod1&resourceVersion=1&watch=true";
+    final String pathWithoutResourceVersion = "/api/v1/namespaces/test/pods?fieldSelector=metadata.name%3Dpod1&watch=true";
+
+    server.expect()
+        .withPath(path)
+        .andReturn(200, new StatusBuilder().withCode(200).build()).once();
+    server.expect()
+        .withPath(path)
+        .andReturn(200, mapper.writeValueAsString(outdatedEvent) + '\n').once();
```
Comment (honnix) [inline kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/WatchTest.java]: I couldn't figure out a better way to do this. The mock server requires a strict line with line feed.

Comment (honnix) [reply kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/WatchTest.java]: Also I'm not entirely sure HTTP code 200 with 410 in the body is really what k8s returns.

**Comment by honnix** [comment]
Now this makes me wondering whether watching from the beginning of history is a bad idea. Can someone point me how `resourceVersion` is set initially before creating a `WatchConnectionManager`?

File: kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/WatchOverHTTP.java [hunk 2]
```diff
@@ -92,55 +92,38 @@ public void onClose(KubernetesClientException cause) {}
   }
 
   @Test
-  public void testOutdated() throws InterruptedException {
-    logger.info("testOutdated");
+  public void testHttpErrorReconnectOutdatedAndModified() throws InterruptedException {
```
Comment (honnix) [inline kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/WatchOverHTTP.java]: This file is not named properly and doesn't seem to be picked up by junit.

**Comment by honnix** [comment]
Maybe it is ok because when this happens, old resource versions got wiped out from history.

**Comment by honnix** [comment]
@rohanKanojia PTAL again. Thanks.

**Comment by rohanKanojia** [comment]
@honnix : Could you please add a line to CHANGELOG regarding this change?

**Comment by honnix** [comment]
@rohanKanojia Sure. I saw that label but didn't realise I should update it. Will do.

**Comment by rohanKanojia** [comment]
[merge]

**Comment by wiggzz** [comment]
This feels like it breaks the contract with the watcher. 

If we get an HTTP_GONE from Kubernetes, then we need to notify the watcher we are going to start again from the beginning - otherwise deletion events can be missed. 

Think about the case where the user of this library has some internal cache of the resources being watched. If we start again from the beginning, this means k8s will send us everything that currently exists, but may not send us deletion events from things that used to exist. 

For instance: we perform a list on all the resources in the cluster and get some current resource version "1234". Then we wait some period of time (say, an hour). In this time, some resources are deleted. Then we ask to see any changes since that point in time by specifying a resource version "1234" with our watch request. If this is too old (and k8s cannot track back that far), then this change means that transparently the watcher will just send us ADDED events for all the resources currently in the cluster, but the deletion events that happened since resource version "1234" will not be sent. This will mean if we were trying to maintain some local cache, the resources that were deleted will be leaked.

If you are going to do this, you'll need to notify the caller to delete everything in their cache because k8s can no longer send all the changes from the resource version they specified.

**Comment by honnix** [comment]
@wiggzz Thanks for the explanation and it is a very good point. I had similar worries when working on this PR because as you pointed out it does break the contract.

I'm fine reverting this change.

**Comment by honnix** [comment]
@wiggzz PTAL https://github.com/fabric8io/kubernetes-client/pull/1853

**Comment by wiggzz** [comment]
Thanks - appreciate the quick turn around.

**Comment by honnix** [comment]
Likewise, thanks for pointing out the defect.

**Comment by rohanKanojia** [comment]
@honnix @wiggzz : Thanks guys, I will cut a release soon this week and let your guys know

**Comment by xiateapu** [comment]
I'm a novice, so.... could you please tell me when or why kubernetes api returns HTTP_GONE?

**Comment by wiggzz** [comment]
> I'm a novice, so.... could you please tell me when or why kubernetes api returns HTTP_GONE?

In the context of a watch, it will return `HTTP_GONE` when you ask to see changes for a `resourceVersion` that is too old - i.e. when it can no longer tell you what has changed since that version, since too many things have changed. In that case, you'll need to start again, by not specifying a `resourceVersion` in which case the watch will send you the current state of the thing you are watching and then send updates from that point.

**Comment by rohanKanojia** [comment]
@xiateapu : ^^

**Comment by xiateapu** [comment]
Actually i had this problem as shown below:  
image

It seems that `resourceVersion` has a big gap between last watched and current ?   
So i'm very confused that why dose the `too many things have changed` happened？Can‘t kubernetes tell me these change in time?

**Comment by xiateapu** [comment]
> > I'm a novice, so.... could you please tell me when or why kubernetes api returns HTTP_GONE?
> 
> In the context of a watch, it will return `HTTP_GONE` when you ask to see changes for a `resourceVersion` that is too old - i.e. when it can no longer tell you what has changed since that version, since too many things have changed. In that case, you'll need to start again, by not specifying a `resourceVersion` in which case the watch will send you the current state of the thing you are watching and then send updates from that point.

Thanks for your answer ^ ^

**Comment by xiateapu** [comment]
> @xiateapu : ^^

^ ^

**Comment by wind57** [comment]
> > I'm a novice, so.... could you please tell me when or why kubernetes api returns HTTP_GONE?
> 
> In the context of a watch, it will return `HTTP_GONE` when you ask to see changes for a `resourceVersion` that is too old - i.e. when it can no longer tell you what has changed since that version, since too many things have changed. In that case, you'll need to start again, by not specifying a `resourceVersion` in which case the watch will send you the current state of the thing you are watching and then send updates from that point.

what about an informer? `ResourceEventHandler` has a default method `onNothing`. Its documentation says the same thing : 

> Called after an empty list is retrieved on start or after an HTTP GONE when the Store is empty

I am asking this in the context of spring-cloud-kubernetes project. Previously, we had used Watches for detecting changes in config maps/secrets. I am trying to move to informers now. So our code is like this at the moment:

```
					@Override
					public void onClose(WatcherException exception) {
						log.warn("ConfigMaps watch closed", exception);
						Optional.ofNullable(exception).map(e -> {
							log.debug("Exception received during watch", e);
							return exception.asClientException();
						}).map(KubernetesClientException::getStatus).map(Status::getCode)
								.filter(c -> c.equals(HttpURLConnection.HTTP_GONE)).ifPresent(c -> watch());
					}
``` 

If I move this to informers, it seems that I should handle this in the same manner, something like:

```
				@Override
				public void onNothing() {
					boolean isStoreEmpty = informer.getStore().list().isEmpty();
					if(!isStoreEmpty) {
						// HTTP_GONE, thus re-inform
						inform();
					}
				}
```

I am a bit lost (due to my limited knowledge) between this should be handled, or an informer (like a watcher) will know what to do on its own.

Can some of the pros comment on this please? thank you.

**Comment by manusa** [comment]
Hi @wind57 
There have been a few issues on this repo about Spring Cloud Kubernetes and the HTTP Gone problem. With more recent versions of the client, this should be less of a problem since watchers are started using [bookmarks](https://kubernetes.io/docs/reference/using-api/api-concepts/#watch-bookmarks) (if the cluster supports it).

In any case, [as suggested](https://github.com/fabric8io/kubernetes-client/issues/2787#issuecomment-1046768554) in some of the reported issues, Spring Cloud should migrate from using raw Watchers to Informers.

Informers are self-healing, so as a client consumer you should not worry about reconnecting, and so on. You can use the `onNothing` event to log or monitor these incidents, but you shouldn't do any error-recovery stuff.

For your use-case, you can consider Informers as improved Watchers which take care of all the stuff for you.

**Comment by wind57** [comment]
thank you @manusa. this is exactly what I am trying to do currently, move to informers, it's [almost ready](https://github.com/spring-cloud/spring-cloud-kubernetes/pull/1036/files#diff-8127d11528a34e5857472226e2ea61b5f90eb00b4038247cf428c30d429ae691) For the time being I will comment `onNothing` and see if we get any reports in the future.

As usual, much appreciate your answer.

**Pull request merged**